specs/interop/liquidity_migration.md

```diff
@@ -0,0 +1,266 @@
+access control function looks like this:
+
+```solidity
+function convert(address _from, address _to, uint256 _amount) public {
+  _validatePair(_from, _to);
+  IERC20(_from).burn(msg.sender, _amount);
+  IERC20(_to).mint(msg.sender, _amount);
+  emit Converted(_from, _to, msg.sender, _amount);
+}
+```
+
+#### `Converted`
+
+The `L2StandardBridge` SHOULD add a `Converted` event
+that MUST trigger when anyone converts tokens
+with `convert`.
+
+```solidity
+event Converted(address indexed from, address indexed to, address indexed caller, uint256 amount);
+```
+
+where `from` is the address of the input token, `to` is the address of the output token,
+`caller` is the `msg.sender` of the function call and `amount` is the converted amount.
+
+#### `_validatePair`
+
+The `L2StandardBridge` SHOULD add a `_validatePair` internal function that
+checks the validity of the `_from` and `_to` pair.
+To do so, it will verify that:
+
+1. Both tokens have the same amount of decimals.
+2. There is one valid legacy address:
+   Verify the `OptimismMintableERC20Factories` deployment lists.
+3. There is one valid `SuperchainERC20` address:
+   Verify the `SuperchainERC20Factory` deployment list.
+4. Both tokens correspond to the same remote token address:
+   it will compare the remote token addresses stored in the deployment
+   mappings and revert with `InvalidPair` if they do not match.
+
+You can read more details about access control in the [Appendix](#access-control).
+
+A reference implementation would look like follows
+
+```solidity
+IOptimismMintableERC20Factory public factory = IOptimismMintableERC20Factory(Predeploys.OptimismMintableERC20Factory);
+ISuperchainERC20Factory public superFactory = ISuperchainERC20Factory(Predeploys.SuperchainERC20Factory);
+
+function _validatePair(address _from, address _to) internal {
+  // 1. Decimals check
+  require(IERC20Metadata(_from).decimals() == IERC20Metadata(_to).decimals(), "Decimals do not match")
+
+  if (_isOptimismMintableERC20(_from)) {
+    _validateFactories(_from, _to)
+  } else {
+    _validateFactories(_to, _from)
+  }
+}
+
+function _validateFactories(address _legacyAddr, address _superAddr) internal {
+  // 2. Valid legacy check
+  address _legacyRemoteToken = factory.deployments(_legacyAddr);
+  require(_legacyRemoteToken != address(0), "Invalid Legacy address");
+
+  // 3. Valid SuperchainERC20 check
+  address _superRemoteToken = superFactory.deployments(_superAddr);
+  require(_superRemoteToken != address(0), "Invalid SuperchainERC20 address");
+
+  // 4. Same remote address check
+  require(_legacyRemoteToken == _superRemoteToken, "Invalid token pair");
+}
+```
+
+Note: Code can be simplified by using || in
+the requires or conditionals.
+The tradeoff would be losing error precision.
+
+### OptimismMintableERC20Factory
+
+#### Version
+
+Ensure the `OptimismMintableERC20Factory` implementation uses CREATE2 instead of CREATE and decimals as part of the salt.
```

Comment:
I think that we should add the spec for how the salt is computed. This is useful so that we can reason about invariants around how the `deployments` mapping is populated.

Right now it is:

```solidity
bytes32 salt = keccak256(abi.encode(_remoteToken, _name, _symbol, _decimals));
```

Since this isn't done with create3, a change in the solidity compiler version will result in different addresses. If it is easy to migrate to a create3 deployer, that would be preferred, but certainly not required. For the new factory we would definitely want to use create3

Reply:
I changed the paragraph to be as open as possible, only discouraging the use of CREATE, discussing the issue with CREATE2 and suggesting CREATE3 eventually. 

Would you like us to work on a dedicated spec for the salt of the `OptimismMintableERC20Factory` then?

Reply:
What is the additional work in your mind besides what i linked above?

Reply:
Nevermind, lol. It would not be worth.
I modified the paragraph in the latest commit. Let me know what do you think of it.